### DISCUSSION TITLE
Where to find and how to install certain extensions

### DISCUSSION THREAD

**Discussion by Sheldon-Monk** [leokhoa/laragon#614] [Q&A]
There is an app (a CRM) which requires the following:
BCMath PHP Extension
Ctype PHP Extension
Fileinfo PHP extension
JSON PHP Extension
Mbstring PHP Extension
OpenSSL PHP Extension
PDO PHP Extension
Tokenizer PHP Extension
XML PHP Extension
MySQL Database
GD Library > or = v2.0

Where do I find those and what is the proper way to install them on Laragon???

Thanks in advance.

**Comment by iamrobert** [comment]
MySQL is already installed.

PHP Extensions:
1. Right Click and follow this path:
![](https://i.imgur.com/wY6EmVz.png)

**Reply by Sheldon-Monk** [reply]
But I'm not finding many on the pop-out list.  Where do we find those and how best to install those when running Laragon on Windows 11???

**Reply by iamrobert** [reply]
They come installed by default:
https://medium.com/ariel-mejia-dev/laragon-setup-series-part-4-add-php-extensions-on-laragon-be0defb6b913

Why don't you share a screenshot of what you see when you go to:

PHP > Extensions

**Reply by iamrobert** [reply]
How to install different PHP versions:
https://pen-y-fan.github.io/2023/01/15/how-to-update-the-php-version-in-laragon/

**Reply by Sheldon-Monk** [reply]
None of that answers my original question.  All the PHP versions I sifted through DO NOT CONTAIN:
BCMath
Ctype
Tokenizer
XML

Also, there are 5 PDO extensions listed by not a single listing of just 'PDO'; therefore no way of telling if those 5 are all in that library.

according to the dropdown menu/PHP/extensions,  So, again, where can I find those and get the instructions to install when Laragon is installed on Win11Pro?

**Reply by iamrobert** [reply]
Windows PHP has full support for all the extensions you mentioned and they are installed by default:

Ctype:
https://www.php.net/manual/en/ctype.installation.php

BCMath
https://www.php.net/manual/en/bc.installation.php

Tokenizer:
https://www.php.net/manual/en/tokenizer.installation.php

XML:
https://www.php.net/manual/en/xml.installation.php

I ran this code under my localhost, and they are installed by default in PHP.

```
<?php

echo extension_loaded("xml") ? "XML is loaded" : "XML is not loaded";
echo extension_loaded("ctype") ? "Ctype is loaded" : "Ctype is not loaded";
echo extension_loaded("tokenizer") ? "Tokenizer is loaded" : "Tokenizer is not loaded";
echo extension_loaded("xml") ? "XML is loaded" : "XML is not loaded";


?>
```

![image](https://github.com/leokhoa/laragon/assets/8617673/a06e8211-9d47-4957-a0c7-707ff4fc9f5e)

**Reply by iamrobert** [reply]
In answer to how do you check in Laragon - its similar to how to check on a webhost via PHP info.


**Check Extensions:**

1. Create a `phpinfo();` file to list loaded extensions. You can also navigate to `http://localhost` and click the **info link** to view PHP info
2. Look for BCMath, Ctype, Tokenizer, XML etc.

![image](https://github.com/leokhoa/laragon/assets/8617673/f40b1fdd-6289-4b30-9f30-0889945e24b6)

**Enable Extensions in php.ini:**
If the extensions are installed but not enabled, you can enable them via the php.ini file. However, the ones you listed are installed by default.

1. In Laragon, open the php.ini file - PHP > PHP.ini.
2. Look for extensions (;extension=name) and remove the semicolon (;) to uncomment and enable them.
3. Restart Apache for the changes to take effect.

**Install Extensions in Windows**
Search for extension_dir in your PHP Info file. This will give you the path to the PHP extensions directory.

![image](https://github.com/leokhoa/laragon/assets/8617673/5a68f8d8-4482-4642-8afa-9b142e0431fe)

- You can install the .dll files in the extension_dir of your Laragon's PHP installation.
- Add the extension to the php.ini
- restart Apache for the changes to take effect.

**Reply by Sheldon-Monk** [reply]
Thanks.  (Been away for a number of months due to medical issues.  I think your hints worked.  Thanks again.

**Reply by Sheldon-Monk** [reply]
Still, it would be good to know where online one can find those add-ons, for future reference.  Thanks again.
